libr/util/idpool.c

```diff
@@ -163,6 +162,71 @@ R_API void* r_id_storage_get(RIDStorage* storage, ut32 id) {
 	return storage->data[id];
 }
 
+R_API bool r_id_storage_get_lowest(RIDStorage *storage, ut32 *id) {
+	r_return_val_if_fail (storage, false);
+	int fd = 0;
+	if (!r_id_storage_get_prev (storage, (ut32 *)&fd) && fd == 0) {
+		return false;
+	}
+	if (!r_id_storage_get_next (storage, (ut32 *)&fd) && fd == 0) {
+		return false;
+	}
+	*id = fd;
+	return true;
+}
+
+R_API bool r_id_storage_get_highest(RIDStorage *storage, ut32 *id) {
+	r_return_val_if_fail (storage, false);
+	int fd = 0;
+	if (!r_id_storage_get_prev (storage, (ut32 *)&fd) && fd == 0) {
+		return false;
+	}
+	*id = fd;
+	return true;
+}
+
+// usually these shall not be used
+// id-storage doesn't provide sorting things in specific order
+// there are only a very few reasonable usecases, such as cycling through RIODescs
+R_API bool r_id_storage_get_next(RIDStorage *storage, ut32 *id) {
+	r_return_val_if_fail (id && storage && storage->data, false);
+	// it does not indicate a bug, if the passed id goes beyond the boundaries of the storage
+	// therefor no r_return_val_if_fail for this case
+	if (storage->size <= *id) {
+		return false;
+	}
+
+	ut32 _id = *id;
+	const ut32 start_id = _id;
+	do {
+		_id = (_id + 1) % storage->size;
+		if (_id == start_id) {
+			break;
+		}
+	} while (!storage->data[_id]);
+	*id = _id;
+	return (_id >= start_id); //true;
+	return true;
+}
+
+R_API bool r_id_storage_get_prev(RIDStorage *storage, ut32 *id) {
+	r_return_val_if_fail (id && storage && storage->data, false);
+	if (storage->size <= *id) {
+		return false;
+	}
+
+	ut32 _id = *id;
+	const ut32 start_id = _id;
+	do {
+		_id = (_id + storage->size - 1) % storage->size;
+		if (_id == start_id) {
+			break;
+		}
+	} while (!storage->data[_id]);
+	*id = _id;
+	return (_id <= start_id); //true;
```

Comment:
this is wrong. should always return true

Reply:
nope. then theres no way to know if we are in the latest or lowest fd. in your previous implementation rotate and next was doing the same for example. and it was never returning false unless there was no elements in the storage.

Reply:
yes, that's intended, so you can rotate forward, like in a ring-buffer. if you have ids 3,4,5 and do id_next for 5 the resulting id should be 3

Reply:
it does that, but returns false, so you can deterine if the ring buffer have looped or not

Reply:
but the return value should indicate an error

Reply:
yes, the error is that there´s no such next or prev fd because you need to cicle back to the lower fds to get the next one and this is what its returned by reference.